## Exercise - DL Tutorial 1 

Please complete the following notebook and submit your solutions to manuel.milling@informatik.uni-augsburg.de

## student name: 

Solutions from exercise sheet 3 (class methods below).

In [18]:
import numpy as np
#numpy random seed
np.random.seed(42)

trainx, trainy, testx, testy = np.load('mnist.npy', allow_pickle=True)
print("Trainx shape: {}".format(trainx.shape))
print("Trainy shape: {}".format(trainy.shape))
print("Testx shape:  {}".format(testx.shape))
print("Testy shape:  {}".format(testy.shape))

def sigmoid(X):
    return 1/(1 +np.exp(-X))

def softmax(X):
    #more stable
    eps = X.max()
    return np.exp(X + eps)/(np.sum(np.exp(X + eps), axis=1).reshape((X.shape[0],1)))

def fcc_one_layer(X, W, b, activation):
    return activation(np.matmul(X, W) + b)

def cross_entropy(pred_logits, y):
    num_data_points = pred_logits.shape[0]
    correct_logits = pred_logits[np.arange(num_data_points),y]
    return np.mean(-np.log(correct_logits))

def accuracy(logits, labels):
    class_predictions = np.argmax(logits, axis=1)
    return np.mean(class_predictions == labels)



Trainx shape: (60000, 784)
Trainy shape: (60000,)
Testx shape:  (10000, 784)
Testy shape:  (10000,)


1.   Implement the error of the last layer.

In [19]:
def delta_last_layer(h, y):
    """
    :param h: softmax activations of shape (num_examples, num_classes)
    :param y: correct labels of shape (num_examples,)
    :return: delta of of last layer, i.e. derivative of cross entropy times derivative of softmax
    """
    pass

2.   Implement the derivative of the sigmoid function in terms of the sigmoid function.

In [20]:
def del_sigmoid(h):
    """
    :param h: output of the sigmoid function shape (num_examples, num_units)
    :return: element-wise derivative of the sigmoid function
    """
    pass


3.   Implement the backpropagation as a class method.
4.   Implement the the optimisation step as a class method.


In [21]:
class fcc:
    def __init__(self, n_input, n_hidden1, n_hidden2, n_out):
        #parameters
        self.W_i_h1 = np.random.randn(n_input, n_hidden1)
        self.b_h1 = np.random.randn(n_hidden1)
        self.W_h1_h2 = np.random.randn(n_hidden1, n_hidden2)
        self.b_h2 = np.random.randn(n_hidden2)
        self.W_h2_o = np.random.randn(n_hidden2, n_out)
        self.b_out = np.random.randn(n_out)
        #neuron activations and input H^n
        self.X = None
        self.h1 = None
        self.h2 = None
        self.out = None
        #components of the gradient
        self.dW_i_h1 = None
        self.db_h1 = None
        self.dW_h1_h2 = None
        self.db_h2 = None
        self.dW_h2_o = None
        self.db_out = None

        n_trainable_bias = self.b_h1.shape[0] + self.b_h2.shape[0] + self.b_out.shape[0]
        n_trainable_weights = self.W_i_h1.shape[0] * self.W_i_h1.shape[1] + self.W_h1_h2.shape[0] * self.W_h1_h2.shape[1] + self.W_h2_o.shape[0] * self.W_h2_o.shape[1]
        print("Number of parameters: {}".format(n_trainable_bias + n_trainable_weights))

    def forward_propagation(self, X):
        self.X = X
        self.h1 = fcc_one_layer(X, self.W_i_h1, self.b_h1, sigmoid)
        self.h2 = fcc_one_layer(self.h1, self.W_h1_h2, self.b_h2, sigmoid)
        self.out = fcc_one_layer(self.h2, self.W_h2_o, self.b_out, softmax)
        return self.out

    def backprop(self, y):
        """
        :param y: labels, i.e. numbers of correct classes of shape (num_examples,)
        """
        pass
    def gradient_step(self, learning_rate):
        """
        :param learning_rate: learning_rate for training
        """
        pass


5.   Implement the training routine.

In [22]:
learning_rate = 0.01
num_iterations = 1000

neural_net = fcc(784, 400, 400, 10)
logits = neural_net.forward_propagation(trainx)
print("Train Loss:\t{}".format(cross_entropy(logits, trainy)))
print("Train Accuracy:\t{}".format(accuracy(logits, trainy)))

Number of parameters: 478410
Train Loss:	20.01954426412725
Train Accuracy:	0.11021666666666667


9.   Which are the parameters we can tune to improve the performance of our network? How many trainable parameters (scalars) does our network have in total?

The tunable parameters are all weights and biases, in total 478410, see calculation above.

10.  Why did we implement two different evaluation metrics of our system (cross-entropy and accuracy)? What are the main differences between the two and why can’t/shouldn’t we use them interchangeably?

A crucial property of the loss function is that it has to be differentiable at (almost) all points. The accuracy, predicting the class with the highest probability is almost never differentiable, as infinitesimal changes to the network do not change the overall prediction. The cross-entropy however is differentiable in almost every point, so small changes in the network impact the loss function of the network.
The accuracy however is the more intuitive and praxis-oriented evaluation metric of a classification problem, as it forces the network to make one prediction. For most applied system it makes more sense to target 100% accuracy, rather than a low probability distribution in false classes. 
